proposals/XXXX-build-time-constant-values.md

```diff
@@ -0,0 +1,188 @@
+
+Attribute definitions can benefit from additional guarantees of compile-time constant values. Imagine a property wrapper that declares a property is to be serialized and that it must be stored/retrieved using a specific string key. `Codable` requires users to provide a `CodingKeys` `enum` boilerplate, relying on the `enum`’s `String` raw values. Alternatively, such key can be specified on the property wrapper itself:
+
+```
+struct Foo {
+  @SpecialSerializationSauce(key: "title") 
+  var someSpecialTitleProperty: String
+}
+
+@PropertyWrapper
+struct SpecialSerializationSauce {
+  init(key: const String) {...}
+}
+```
+
+Having the compiler enforce the compile-time constant property of `key` parameter eliminates the possibility of an error where a run-time value is specified which can cause serialized data to not be able to be deserialized, for example. 
+
+Enforcing compile-time constant nature of the parameters is also the first step to allowing attribute/library authors to be able to check uses by performing compile-time sanity checking and having the capability to emit custom build-time error messages.
+
+### Enforcement of Non-Failable Initializers 
+
+Ergonomics of the recently-pitched `Foundation.URL` [insert link here] would benefit greatly from the ability to require the string argument to be compile-time constant. With evolving compile-time evaluation facilities, Swift may even gain an ability to perform compile-time validation of such URLs  even though the user may never be able to express a fully compile-time constant `Foundation.URL` type because this type is a part of an ABI-stable SDK. While a type like `StaticString` may be used to require that the argument string must be static, which string is chosen can still be determined at runtime, e.g.:
+
+```
+URL(Bool.random() ? "https://valid.url.com" : "invalid url . com")
+```
+
+### Guaranteed Optimization Hints
+
+Similarly, ergonomics of numeric intrinsics can benefit from allowing only certain function parameters to be required to be compile-time known. For example, requiring a given numeric operation to specify a `@const` parameter for the rounding mode of an operation as an enum case, while allowing the operands of the operation be runtime values, allowing the compiler to generate more-efficient code. 
+
+## Source compatibility
+
+This is a purely additive change and has no source compatibility impacts.
+
+## Effect on ABI stability and API resilience
+
+The new function parameter attribute is a part of name mangling.
+
+## Effect on SwiftPM packages
+
+There is no impact on SwiftPM packages.
+
+## Future Directions
+
+* Compile-time types - allow types consisting of `const` properties to be treated as compile-time-known values.
+* Compile-time expressions - allow expressions that operate on compile-time-known values, from binary operators to control flow.
+* Compile-time functions - consisting of expressions operating on compile-time-known values and `const` parameters.
+
```

Comment:
I would add "toolchain support for extracting compile-time values at compile time"

Reply:
Also we should mention propagation of constants.